# Enforce consistent Zod array style (`zod-x/array-style`)

üíº This rule is enabled in the ‚úÖ `recommended` config.

üîß This rule is automatically fixable by the [`--fix` CLI option](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix).

<!-- end auto-generated rule header -->

## Options

<!-- begin auto-generated rule options list -->

| Name    | Description                                | Type   | Choices              |
| :------ | :----------------------------------------- | :----- | :------------------- |
| `style` | Decides which style for zod array function | String | `function`, `method` |

<!-- end auto-generated rule options list -->

This rule enforces a consistent style for defining arrays in Zod schemas. You can choose between:

- **Function style**: `z.array(...)`
- **Method style**: `.array()`

By default, `function` is used to maintain consistency with TypeScript‚Äôs `Array` generic syntax.

## Examples

### `function`

#### ‚úÖ Valid

```ts
z.array(z.string());
```

#### ‚ùå Invalid

```ts
z.string().array();

z.string().trim().array();
```

### `method`

#### ‚úÖ Valid

```ts
z.string().array(); // method
```

#### ‚ùå Invalid

```ts
z.array(z.string());

z.array(z.string().trim());
```

## Further Reading

- [Array Types in TypeScript](https://tkdodo.eu/blog/array-types-in-type-script)
